web/src/attachments_ui.ts

```diff
@@ -14,31 +15,63 @@ import * as scroll_util from "./scroll_util";
 import * as timerender from "./timerender";
 import * as ui_report from "./ui_report";
 
-let attachments;
-let upload_space_used;
-
-export function bytes_to_size(bytes, kb_with_1024_bytes = false) {
+type ServerAttachment = z.infer<typeof attachment_api_response_schema>["attachments"][number];
+
+type Attachment = ServerAttachment & {
+    create_time_str: string;
+    size_str: string;
+};
+
+type AttachmentEvent = {
+    op: "add" | "remove" | "update";
+    attachment: ServerAttachment;
+    upload_space_used: number;
+};
+
+const attachment_api_response_schema = z.object({
+    attachments: z.array(
+        z.object({
+            id: z.number(),
+            name: z.string(),
+            path_id: z.string(),
+            size: z.number(),
+            create_time: z.number(),
+            messages: z.array(
+                z.object({
+                    id: z.number(),
+                    date_sent: z.number(),
+                }),
+            ),
+        }),
+    ),
+    upload_space_used: z.number(),
+});
+
+let attachments: Attachment[];
+let upload_space_used: z.infer<typeof attachment_api_response_schema>["upload_space_used"];
+
+export function bytes_to_size(bytes: number, kb_with_1024_bytes = false): string {
     const kb_size = kb_with_1024_bytes ? 1024 : 1000;
     const sizes = ["B", "KB", "MB", "GB", "TB"];
     if (bytes === 0) {
         return "0 B";
     }
-    const i = Number.parseInt(Math.floor(Math.log(bytes) / Math.log(kb_size)), 10);
+    const i = Number.parseInt(Math.floor(Math.log(bytes) / Math.log(kb_size)).toString(), 10);
```

Comment:
The original code seems to be [incorrect](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#using_parseint_on_non-strings). Let's maybe do this instead:

```ts
    const i = Math.trunc(Math.log(bytes) / Math.log(kb_size));
```